### ISSUE TITLE
动态图测试精度随batchsize和sampler顺序的变动而变动

### ISSUE THREAD

**Issue by flytocc** [PaddlePaddle/Paddle#42429]
版本、环境信息：
   1）PaddlePaddle版本：2.2.2
   2）GPU：V100 (3090 及 TiTAN Xp也存在此问题)
   3）CUDA： 10.2
   4）系统环境：python 3.7，AI Studio

AI Studio项目地址：https://aistudio.baidu.com/aistudio/projectdetail/3926029

model (save as models.py)如下：
```python
import paddle
import paddle.nn as nn


trunc_normal_ = nn.initializer.TruncatedNormal(std=.02)
zeros_ = nn.initializer.Constant(value=0)


def drop_path(x, drop_prob=0., training=False):
    """Drop paths (Stochastic Depth) per sample (when applied in main path of residual blocks).
    the original name is misleading as 'Drop Connect' is a different form of dropout in a separate paper...
    See discussion: https://github.com/tensorflow/tpu/issues/494#issuecomment-532968956 ...
    """
    if drop_prob == 0. or not training:
        return x
    keep_prob = paddle.to_tensor(1 - drop_prob)
    shape = (paddle.shape(x)[0], ) + (1, ) * (x.ndim - 1)
    random_tensor = keep_prob + paddle.rand(shape, dtype=x.dtype)
    random_tensor = paddle.floor(random_tensor)  # binarize
    output = x.divide(keep_prob) * random_tensor
    return output


class DropPath(nn.Layer):
    """Drop paths (Stochastic Depth) per sample  (when applied in main path of residual blocks).
    """

    def __init__(self, drop_prob=None):
        super(DropPath, self).__init__()
        self.drop_prob = drop_prob

    def forward(self, x):
        return drop_path(x, self.drop_prob, self.training)


class Mlp(nn.Layer):
    def __init__(self,
                 in_features,
                 hidden_features=None,
                 out_features=None,
                 act_layer=nn.GELU,
                 drop=0.):
        super().__init__()
        out_features = out_features or in_features
        hidden_features = hidden_features or in_features
        self.fc1 = nn.Linear(in_features, hidden_features)
        self.act = act_layer()
        self.fc2 = nn.Linear(hidden_features, out_features)
        self.drop = nn.Dropout(drop)

    def forward(self, x):
        x = self.fc1(x)
        x = self.act(x)
        x = self.drop(x)
        x = self.fc2(x)
        x = self.drop(x)
        return x


class Attention(nn.Layer):
    def __init__(self,
                 dim,
                 num_heads=8,
                 qkv_bias=False,
                 qk_scale=None,
                 attn_drop=0.,
                 proj_drop=0.):
        super().__init__()
        self.num_heads = num_heads
        head_dim = dim // num_heads
        self.scale = qk_scale or head_dim**-0.5

        self.qkv = nn.Linear(dim, dim * 3, bias_attr=qkv_bias)
        self.attn_drop = nn.Dropout(attn_drop)
        self.proj = nn.Linear(dim, dim)
        self.proj_drop = nn.Dropout(proj_drop)

    def forward(self, x):
        # B= paddle.shape(x)[0]
        N, C = x.shape[1:]
        qkv = self.qkv(x).reshape(
            [-1, N, 3, self.num_heads, C // self.num_heads]).transpose([2, 0, 3, 1, 4])
        q, k, v = qkv[0], qkv[1], qkv[2]

        attn = (q * self.scale) @ k.transpose([0, 1, 3, 2])
        attn = nn.functional.softmax(attn, axis=-1)
        attn = self.attn_drop(attn)

        x = (attn @ v).transpose([0, 2, 1, 3]).reshape([-1, N, C])
        x = self.proj(x)
        x = self.proj_drop(x)
        return x


class Block(nn.Layer):
    def __init__(self,
                 dim,
                 num_heads,
                 mlp_ratio=4.,
                 qkv_bias=False,
                 qk_scale=None,
                 drop=0.,
                 attn_drop=0.,
                 drop_path=0.,
                 act_layer=nn.GELU,
                 norm_layer=nn.LayerNorm,
                 skip_lam=1.):
        super().__init__()
        self.skip_lam = skip_lam

        self.norm1 = norm_layer(dim)
        self.attn = Attention(
            dim,
            num_heads=num_heads,
            qkv_bias=qkv_bias,
            qk_scale=qk_scale,
            attn_drop=attn_drop,
            proj_drop=drop)
        self.drop_path = DropPath(
            drop_path) if drop_path > 0. else nn.Identity()
        self.norm2 = norm_layer(dim)
        mlp_hidden_dim = int(dim * mlp_ratio)
        self.mlp = Mlp(in_features=dim,
                       hidden_features=mlp_hidden_dim,
                       act_layer=act_layer,
                       drop=drop)

    def forward(self, x):
        x = x + self.drop_path(self.attn(self.norm1(x))) / self.skip_lam
        x = x + self.drop_path(self.mlp(self.norm2(x))) / self.skip_lam
        return x


class PatchEmbed4_2(nn.Layer):
    """ 
    Image to Patch Embedding with 4 layer convolution
    """

    def __init__(self,
                 img_size=224,
                 patch_size=16,
                 in_chans=3,
                 embed_dim=768):
        super().__init__()

        self.num_patches = (img_size // patch_size)**2

        self.conv1 = nn.Conv2D(
            in_chans, 64, 7, stride=2, padding=3, bias_attr=False)  # 112x112
        self.bn1 = nn.BatchNorm2D(64)

        self.conv2 = nn.Conv2D(
            64, 64, 3, stride=1, padding=1, bias_attr=False)  # 112x112
        self.bn2 = nn.BatchNorm2D(64)

        self.conv3 = nn.Conv2D(
            64, 64, 3, stride=1, padding=1, bias_attr=False)
        self.bn3 = nn.BatchNorm2D(64)

        self.relu = nn.ReLU()
        self.proj = nn.Conv2D(
            64, embed_dim, patch_size // 2, stride=patch_size // 2)

    def forward(self, x):
        x = self.conv1(x)
        x = self.bn1(x)
        x = self.relu(x)

        x = self.conv2(x)
        x = self.bn2(x)
        x = self.relu(x)

        x = self.conv3(x)
        x = self.bn3(x)
        x = self.relu(x)

        x = self.proj(x)  # [B, C, W, H]

        return x


class LV_ViT(nn.Layer):
    """ Vision Transformer with tricks
    Arguements:
        skip_lam: residual scalar for skip connection (default: 1.0)
    """

    def __init__(self, img_size=224, patch_size=16, in_chans=3, num_classes=1000, embed_dim=768,
                 depth=12, num_heads=12, mlp_ratio=4., qkv_bias=False, qk_scale=None, drop_rate=0.,
                 attn_drop_rate=0., drop_path_rate=0., norm_layer=nn.LayerNorm, skip_lam=1.0):
        super().__init__()
        self.patch_embed = PatchEmbed4_2(
            img_size=img_size, patch_size=patch_size, in_chans=in_chans, embed_dim=embed_dim)
        num_patches = self.patch_embed.num_patches

        self.cls_token = self.create_parameter(
            shape=[1, 1, embed_dim])
        self.pos_embed = self.create_parameter(
            shape=[1, num_patches + 1, embed_dim])
        self.pos_drop = nn.Dropout(p=drop_rate)

        dpr = [x.item() for x in paddle.linspace(0, drop_path_rate, depth)]
        self.blocks = nn.LayerList([
            Block(dim=embed_dim, num_heads=num_heads, mlp_ratio=mlp_ratio, qkv_bias=qkv_bias,
                  qk_scale=qk_scale, drop=drop_rate, attn_drop=attn_drop_rate, drop_path=dpr[i],
                  norm_layer=norm_layer, skip_lam=skip_lam)
            for i in range(depth)])
        self.norm = norm_layer(embed_dim)

        self.head = nn.Linear(
            embed_dim, num_classes) if num_classes > 0 else nn.Identity()
        self.aux_head = nn.Linear(
            embed_dim, num_classes) if num_classes > 0 else nn.Identity()

        trunc_normal_(self.pos_embed)
        trunc_normal_(self.cls_token)
        self.apply(self._init_weights)

    def _init_weights(self, m):
        if isinstance(m, nn.Linear):
            trunc_normal_(m.weight)
            if m.bias is not None:
                zeros_(m.bias)

    def forward_tokens(self, x):
        B = x.shape[0]
        cls_tokens = self.cls_token.expand([B, -1, -1])
        x = paddle.concat([cls_tokens, x], axis=1)
        x = x + self.pos_embed
        x = self.pos_drop(x)
        for blk in self.blocks:
            x = blk(x)
        x = self.norm(x)
        return x

    def forward(self, x):
        x = self.patch_embed(x)
        x = x.flatten(2).transpose([0, 2, 1])
        x = self.forward_tokens(x)

        x_cls = self.head(x[:, 0])
        x_aux = self.aux_head(x[:, 1:])
        return x_cls + 0.5 * x_aux.max(1)[0]


def lvvit_t(**kwargs):
    model = LV_ViT(patch_size=16,
                   embed_dim=240,
                   depth=12,
                   num_heads=4,
                   mlp_ratio=3.,
                   skip_lam=1.,
                   **kwargs)
    return model
```

测试脚本(save as eval.py)如下：
```python
import argparse
import os
import tqdm

import paddle
import paddle.io
import paddle.nn as nn
import paddle.vision.datasets as datasets
import paddle.vision.transforms as transforms

import models  # lvvit_t


def get_args_parser():
    parser = argparse.ArgumentParser(
        'TokenLabeling evaluation script', add_help=False)
    parser.add_argument('--batch_size', default=64, type=int,
                        help='Batch size per GPU')
    parser.add_argument('--data_path', default='/datasets/imagenet/', type=str,
                        help='dataset path')
    parser.add_argument('--finetune', default='',
                        help='finetune from checkpoint')
    parser.add_argument('--shuffle', action='store_true')
    return parser


def accuracy(output, target, topk=(1,)):
    """Computes the accuracy over the k top predictions for the specified values of k"""
    maxk = min(max(topk), output.shape[1])
    batch_size = target.shape[0]
    _, pred = output.topk(maxk, 1, True, True)
    pred = pred.t()
    correct = pred == target.reshape([1, -1]).expand_as(pred)
    return [correct[:min(k, maxk)].reshape([-1]).astype('float32').sum(0) * 100. / batch_size for k in topk]


@paddle.no_grad()
def evaluate(data_loader, model):
    model.eval()

    outputs, targets = [], []
    for images, target in tqdm.tqdm(data_loader):
        output = model(images)
        outputs.append(output)
        targets.append(target)
    outputs = paddle.concat(outputs)
    targets = paddle.concat(targets)

    loss = nn.CrossEntropyLoss()(outputs, targets)
    acc1, acc5 = accuracy(outputs, targets, topk=(1, 5))

    print(f'* Acc@1 {acc1.item():.3f} Acc@5 {acc5.item():.3f} loss {loss.item():.4f}')


def main(args):
    root = os.path.join(args.data_path, 'val')
    transform = transforms.Compose([
        transforms.Resize(int(224 / 0.9), interpolation='bicubic'),
        transforms.CenterCrop(224),
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225]),
    ])
    dataset_val = datasets.DatasetFolder(
        root, transform=transform)
    sampler_val = paddle.io.BatchSampler(
        dataset=dataset_val, batch_size=args.batch_size, shuffle=args.shuffle)
    data_loader_val = paddle.io.DataLoader(
        dataset_val, batch_sampler=sampler_val, num_workers=10)

    model: nn.Layer = models.__dict__['lvvit_t']()

    def freeze_bn(layer):
        if isinstance(layer, (nn.layer.norm._BatchNormBase)):
            layer._use_global_stats = True
    model.apply(freeze_bn)

    if args.finetune:
        checkpoint = paddle.load(args.finetune)
        model.set_state_dict(checkpoint['model'])
        print("Load checkpoint %s" % args.finetune)

    evaluate(data_loader_val, model)


if __name__ == '__main__':
    parser = argparse.ArgumentParser(
        'TokenLabeling evaluation script', parents=[get_args_parser()])
    args = parser.parse_args()
    main(args)
```

测试命令如下：
```bash
python eval.py --data_path /path/to/imagenet/ --finetune lvvit_t_ema.pd --batch_size 128 --shuffle
```

不同batchsize和sampler顺序下的精度：
| batchsize | shuffle | acc1 |
| ---------- | ------- | ---- |
| 64 | x | 82.292 |
| 64 | v | 79.494 |
| 128 | v | 79.532 |
| 256 | v | 79.548 |

可能的原因：
nn.BatchNorm2D可能有问题，但不知是我使用的问题，还是一个bug。

**Comment by windstamp** [comment]
AI Studio项目地址：https://aistudio.baidu.com/aistudio/projectdetail/3926029 , 进去运行发现少了文件 `./data/lvvit_t_ema.pd`，跑不起来。

请问表格中 `shuffle` 的取值 `x` 和 `v` 分别代表什么意思。

正常的测试精度应该是多少，以及测试精度在不同的 batchsize 和 sampler 顺序下正常应该是一致的，这是和哪个参考标准对比的吗。

不排除 `nn.BatchNorm2D` 有问题，但是建议可以先尝试再检查一下模型层面。

**Comment by flytocc** [comment]
> AI Studio项目地址：https://aistudio.baidu.com/aistudio/projectdetail/3926029 , 进去运行发现少了文件 `./data/lvvit_t_ema.pd`，跑不起来。
> 
> 请问表格中 `shuffle` 的取值 `x` 和 `v` 分别代表什么意思。
> 
> 正常的测试精度应该是多少，以及测试精度在不同的 batchsize 和 sampler 顺序下正常应该是一致的，这是和哪个参考标准对比的吗。
> 
> 不排除 `nn.BatchNorm2D` 有问题，但是建议可以先尝试再检查一下模型层面。

我更新了AI Studio项目，现在应该有`./work/lvvit_t_ema.pd`这个文件了。或者从[百度网盘](https://pan.baidu.com/s/1UCU9zTrRo021lDbd60scpA?pwd=vpeg)下载。

 `shuffle` 的取值 `x` 和 `v` 分别代表`BatchSampler`中shuffle=False和shuffle=True。

通过转到pytorch测试出精度是`79.638`，在pytorch中不存在batchsize和shuffle影响精度的问题。

我觉得模型没什么问题，只有Conv2D、Linear、LayerNorm、BatchNorm2D、DropPath、Dropout、Softmax，没有什么特殊的操作。和[ViT](https://github.com/PaddlePaddle/PaddleClas/blob/release/2.3/ppcls/arch/backbone/model_zoo/vision_transformer.py)相比，不同之处在于`PatchEmbed4_2`中的`Conv2D`和`BatchNorm2D`。在pytorch中没出现任何问题。

**Comment by windstamp** [comment]
> 通过转到pytorch测试出精度是79.638，在pytorch中不存在batchsize和shuffle影响精度的问题。

那是否可以理解，现在的问题是 batchsize=64，shuffle=False，acc1=82.292 精度出现了问题。上述表格中其它取值时精度是正常的。还是说在 pytorch 中上述取值全部是 79.638。

> 和[ViT](https://github.com/PaddlePaddle/PaddleClas/blob/release/2.3/ppcls/arch/backbone/model_zoo/vision_transformer.py)相比，不同之处在于PatchEmbed4_2中的Conv2D和BatchNorm2D。在pytorch中没出现任何问题。

请问有用 ViT 这个模型代码测试上述参数取值时的精度吗。

**Comment by flytocc** [comment]
> 那是否可以理解，现在的问题是 batchsize=64，shuffle=False，acc1=82.292 精度出现了问题。上述表格中其它取值时精度是正常的。

在paddle中，当shuffle=True（也就是sampler顺序固定）且batchsize固定时，测得的精度是固定的。所以，表格中所给精度之间的误差，都是某种系统性的误差，都是不该存在的。我认为，在paddle上测得的所有精度都是不可信的、不正常的。

> 还是说在 pytorch 中上述取值全部是 79.638。

在pytorch中，我多次测试了`batchsize=64/128/256`，`shuffle=False/True`的精度，结果都是`79.638`。所以，我说：“在pytorch中不存在batchsize和shuffle影响精度的问题。“

> 请问有用 ViT 这个模型代码测试上述参数取值时的精度吗。

我没用Paddle测过[ViT](https://github.com/PaddlePaddle/PaddleClas/blob/release/2.3/ppcls/arch/backbone/model_zoo/vision_transformer.py)这个模型是否存在这个问题。

**Comment by HydrogenSulfate** [comment]
你好，lv_vit模型的forward函数中，`return x_cls + 0.5 * x_aux.max(1)[0]`这一行调用了max这个API，[torch.max](https://pytorch.org/docs/stable/generated/torch.max.html#torch.max)会返回两个值max_value, max_indices，[paddle.max](https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/max_cn.html#fanhui)只会返回max_value。
建议将`LV_ViT.forward`函数修改成以下代码：
```python
def forward(self, x):
    x = self.patch_embed(x)
    x = x.flatten(2).transpose([0, 2, 1])
    x = self.forward_tokens(x)

    x_cls = self.head(x[:, 0])
    x_aux = self.aux_head(x[:, 1:])
    # return x_cls + 0.5 * x_aux.max(1)[0] # torch写法
    return x_cls + 0.5 * x_aux.max(1)  # paddle写法
```
修改完毕之后bs 64/128/256+有无shuffle的情况下，结果全部都是`Acc@1 79.636 Acc@5 94.848 loss 0.8795`。

**Comment by flytocc** [comment]
我已经修正了这个错误。这个issue也在几天前就关闭了。辛苦你了。
